### PR TITLE
[GEM] GEM packer-unpacker update for GEMChMap

### PR DESCRIPTION
**Pull request by yeckang**
#### PR description:
* Current GEM packer and unpacker are using GEMeMap which is not corresponding with GE21 readout system.
* GEM packer and unpacker are updated to use GEMChMap.
* Backports to `CMSSW_12_4_X` and `CMSSW_12_3_X` are needed.

#### PR validation:
* The code format has applied with `scram build code-format` and `scram build code-checks`
* Currently the branch is not compatible with `runTheMatrix.py` command, because GEMChMap is not in the condition that is used for the `runTheMatrix.py`.

### CODE REVIEW

**Comment by tvami** [comment]
urgent

File: DQM/GEM/interface/GEMDAQStatusSource.h [hunk 1]
```diff
@@ -70,9 +69,7 @@ class GEMDAQStatusSource : public GEMDQMBase {
   void SetLabelOHStatus(MonitorElement *h2Status);
   void SetLabelVFATStatus(MonitorElement *h2Status);
 
-  edm::ESGetToken<GEMeMap, GEMeMapRcd> gemEMapToken_;
-  //std::shared_ptr<GEMROMapping> gemROMap_;
-  const GEMeMap *gemEMap_;
+  edm::ESGetToken<GEMChMap, GEMChMapRcd> gemChMapToken_;
```
Comment (tvami) [inline DQM/GEM/interface/GEMDAQStatusSource.h:72]: this could be `const `

File: DQM/GEM/plugins/GEMDAQStatusSource.cc [hunk 2]
```diff
@@ -27,36 +27,36 @@ void GEMDAQStatusSource::fillDescriptions(edm::ConfigurationDescriptions &descri
...
+
+      mapFEDIdToRe_[fedId] = gemChId.region();
+      mapAMC13ToListChamber_[fedId].push_back(gemChId);
+      mapAMCToListChamber_[{fedId, amcNum}].push_back(gemChId);
     }
 
-    gemEMap.reset();
   } else {
     // no EMap in DB, using dummy
     // FIXME: How to add mapFEDIdToRe_ and mapDetIdToAMC_??
```
Comment (tvami) [inline DQM/GEM/plugins/GEMDAQStatusSource.cc:47]: please update the comment, now you actually access `mapFEDIdToRe_` and `mapDetIdToAMC_`

**Comment by yeckang** [comment]
> @yeckang can you please add to the PR description that backports are needed ?

I just updated the PR description. As far as I understood, the backport to 12_4_X is needed, right?

**Comment by tvami** [comment]
> As far as I understood, the backport to 12_4_X is needed, right?

Based on the latest conversation with @jshlee actually back to 12_3_X even

**Comment by makortel** [comment]
Not caused by this PR, but just to note that strictly speaking all this
https://github.com/cms-sw/cmssw/blob/6da9b8ce0b9cc0067ac18d6db197ae7277d1e806/DataFormats/GEMDigi/interface/GEMAMC.h#L153
(i.e. initializing one member of a union and reading from another) is undefined behavior (AFAIK).

C++20 provides `std::bit_cast` to reinterpret the bit patterns of same-size objects. While waiting for us to move to C++20 we have
https://github.com/cms-sw/cmssw/blob/6da9b8ce0b9cc0067ac18d6db197ae7277d1e806/FWCore/Utilities/interface/bit_cast.h#L26-L29

(I don't suggest to address this in this PR, should I open an issue?)

**Comment by tvami** [comment]
type gem

**Comment by francescobrivio** [comment]
> > Backports to CMSSW_12_4_X and CMSSW_12_3_X are needed.
> 
> I thought we are not pushing for backports by this Friday, in order to not jeopardize the release at last moment. Did I misunderstand the ORP meeting yesterday?

@jpata you are correct, this is not intended for this Friday! 
But we will be using 12_3_X till ~mid July, so we can try to have this included sometime later on in June.
We also need to provide GTs to test this (and we are waiting for the correct tags from GEM group in order to create the GTs).

**Comment by jshlee** [comment]
please test

**Comment by tvami** [comment]
@cmsbuild , please test with https://github.com/cms-sw/cmssw/pull/38090

**Comment by tvami** [comment]
Hi @yeckang @jshlee ok so this error above is because the code is nowhere specifying that it's to be used for Run-3 alone, and so it fails for anything before Run-3. Please add the GEM unpacker sequence to the `run3_common` modifier only.

**Comment by tvami** [comment]
I looked into the sequence a bit more, I see you explicitly have an channel map "on" for 2017
https://github.com/cms-sw/cmssw/blob/master/EventFilter/GEMRawToDigi/python/gemPacker_cfi.py#L9

At that point in time, I was not following GEM related things, was there any GEM installed in 2017? How is the channel map different from what we have now in Run-3? If you do want to run this in 2017 MC, @hyunyong please provide another payload for that scenario (I assume that you did have something different in 2017 than what we have now)

**Comment by hyunyong** [comment]
> I looked into the sequence a bit more, I see you explicitly have an channel map "on" for 2017 https://github.com/cms-sw/cmssw/blob/master/EventFilter/GEMRawToDigi/python/gemPacker_cfi.py#L9
> 
> At that point in time, I was not following GEM related things, was there any GEM installed in 2017? How is the channel map different from what we have now in Run-3? If you do want to run this in 2017 MC, @hyunyong please provide another payload for that scenario (I assume that you did have something different in 2017 than what we have now). Otherwise, if you dont need it, just change that line to `False`.
> 
> What about 2018? Is that the same as 2017?

@tvami Yes, both 2017 and 2018 have the GE1/1 slice test chambers.
I don't know the status of the run2 slice test mapping.
They have changed the configuration during Run2, so probably we need two payloads for it.
@jshlee can answer this.

**Comment by jshlee** [comment]
Hi @tvami , for run2 since we only had slice test chambers, so we didn't implement gem in mc. 
For data, the electronics are quite different and I don't think this update can support that anymore.
I think it's best to turn off gem for run2 for now.

**Comment by tvami** [comment]
Hi @jshlee I'm not really sure how to interpret what you are saying.

> so we didn't implement gem in mc.

The failing wfs in https://github.com/cms-sw/cmssw/pull/38062#issuecomment-1137826614
are mostly for MC. So there is a digitization step for GEM in Run-2 MC implemented.

> For data, the electronics are quite different and I don't think this update can support that anymore.

Uhm, that's not great, IIUC you are saying that after this PR, we wont be able to reproduce the reconstruction of Run-2 data?

> I think it's best to turn off gem for run2 for now.

That certainly will resolve the issues in this PR. Is it possible to write the code so that you have whatever you had so far for Run-2, and this new development with the new record for Run-3 (and beyond)?

**Comment by jshlee** [comment]
Hi @tvami , I had a chat with Yechan and we can make this PR backward compatible for run2.

**Comment by tvami** [comment]
Thanks @jshlee ! (please also dont forget to do `code-checks` before submitting the next commit)

**Comment by tvami** [comment]
@jshlee @yeckang do you have an update on this?

**Comment by tvami** [comment]
@yeckang please do code checks and commit it's results

**Comment by yeckang** [comment]
> @yeckang please do code checks and commit it's results

When I tested `scram b code-checks` on the branch, it doesn't change anything.
It seems like the conflict with the other PR.
Let me resolve the conflicts...

**Comment by tvami** [comment]
@cmsbuild , please test with https://github.com/cms-sw/cmssw/pull/38090

**Comment by tvami** [comment]
I dont actually see the https://github.com/cms-sw/cmssw/pull/38062/commits/0f10cc63f058dfa87a637d3232d47b2dfbfd5262 in the final file changes, maybe your last commit where you "resolved conflicts" got rid of that change?

**Comment by yeckang** [comment]
> I dont actually see the [0f10cc6](https://github.com/cms-sw/cmssw/commit/0f10cc63f058dfa87a637d3232d47b2dfbfd5262) in the final file changes, maybe your last commit where you "resolved conflicts" got rid of that change?

I think the last commit (resolving conflict) absorbed the commit...

**Comment by tvami** [comment]
@yeckang ok, so will you push again? As you can see with your latest commit we are back to where we were earlier and the tests fail for <Run3 wfs

**Comment by yeckang** [comment]
> @yeckang ok, so will you push again? As you can see with your latest commit we are back to where we were earlier and the tests fail for <Run3 wfs

Sorry, I couldn't get your words. Do you want to revert the change that I resolve the conflict?

**Comment by tvami** [comment]
To be honest, I'm a bit lost on what's the GEM group is doing...
What's this PR https://github.com/cms-sw/cmssw/pull/38119 ?
Is that expected to work w/o this PR? And w/o the GTs? Is that PR supposed to be tested together with this PR? 

Also @hyunyong created new paylaod requested here
https://cms-talk.web.cern.ch/t/new-gem-electronics-mapping/10914/4
 
I was under the impression that I need to create new GTs for Run-2 as well, add that to my PR about the GTs, test it with this PR. But then I see 38119 which changes the Run-2 modifier

**Comment by jshlee** [comment]
@tvami - sorry for the confusion, let me try to clarify. 
With this PR, we want to be able to read the GE2/1 slice test chambers and the new HW dataformat that will be implemented once this PR is merged.
However, with this PR, we still need to update the python configs to enable GE2/1 slice test chambers.
But we don't want to make rechits with GE2/1, since these rechits will be used in muon reconstruction.
So, we made PR https://github.com/cms-sw/cmssw/pull/38119 to unpack GE2/1 slice test digis, but not make rechits with GE2/1. 
We want to unpack GE2/1 slice test digis so that it can be monitored with onlineDQM.
I hope this clears it up. 
Please let me know if you want more details.

**Comment by tvami** [comment]
@jshlee
ok, so action items
1) I make new GTs to include Run-2 MC and the updated payloads for Run-3 MC/data, update https://github.com/cms-sw/cmssw/pull/38090, we test this PR with 38090
2) then we test https://github.com/cms-sw/cmssw/pull/38119 with this PR 38062 and the 38090 together
correct?

**Comment by jshlee** [comment]
> @jshlee ok, so action items
> 
> 1. I make new GTs to include Run-2 MC and the updated payloads for Run-3 MC/data, update [Add new GEM electronics map to Run-3 MC and data #38090](https://github.com/cms-sw/cmssw/pull/38090), we test this PR with 38090
> 2. then we test [Turn on GEMDigis and turn off GEMRecHits for GE2/1 demonstrator #38119](https://github.com/cms-sw/cmssw/pull/38119) with this PR 38062 and the 38090 together
>    correct?

Yes!

**Comment by tvami** [comment]
@cmsbuild , please test with https://github.com/cms-sw/cmssw/pull/38090

**Comment by tvami** [comment]
@cmsbuild , please abort

**Comment by tvami** [comment]
@cmsbuild , please test with https://github.com/cms-sw/cmssw/pull/38090

**Comment by tvami** [comment]
@cmsbuild , please test with https://github.com/cms-sw/cmssw/pull/38090

**Comment by tvami** [comment]
@cmsbuild , please test with https://github.com/cms-sw/cmssw/pull/38090

**Comment by pmandrik** [comment]
waiting for backports to CMSSW_12_3_X in order to test at DQM p5 playback

**Comment by tvami** [comment]
@pmandrik the usual way of development is that when the master PR is merged we do the backport. I know that for DQM operations that's the other way around.... the backports should definitely not be opened until we all agreed that this is the final code base. So for now I'd suggest the @cms-sw/dqm-l2 team to just sign this PR based on reading the code, and after the IB tests are fine and the backport is submitted, you can do the tests on the P5 machine.  Please note you'll also need the https://github.com/cms-sw/cmssw/pull/38119 in the backport test. Maybe for the backport these two PRs could be merged. 

@cms-sw/orp-l2 please confirm or comment on the strategy for these developments.

**Comment by yeckang** [comment]
> +1
> 
> * AlCaDB part of the change is now all good
> * tests pass, diffs show known problems
> * @yeckang maybe you could squash the commits into one? Thanks!

@tvami After squashing the commits, we will do the test again. Am I right?

**Comment by pmandrik** [comment]
> @pmandrik the usual way of development is that when the master PR is merged we do the backport. I know that for DQM operations that's the other way around.... the backports should definitely not be opened until we all agreed that this is the final code base. So for now I'd suggest the @cms-sw/dqm-l2 team to just sign this PR based on reading the code, and after the IB tests are fine and the backport is submitted, you can do the tests on the P5 machine. Please note you'll also need the #38119 in the backport test. Maybe for the backport these two PRs could be merged.
> 
> @cms-sw/orp-l2 please confirm or comment on the strategy for these developments.

Hi, it is somehow illogical to perform tests after commits are merged, but for sure if it is not a final version of the code then it is better to wait for the backports.

**Comment by tvami** [comment]
> @tvami After squashing the commits, we will do the test again. Am I right?

That's correct yes, but in the meantime @cms-sw/reconstruction-l2 colleagues can have a look at the results here https://github.com/cms-sw/cmssw/pull/38062#issuecomment-1143156517

**Comment by tvami** [comment]
> Hi, it is somehow illogical to perform tests after commits are merged, but for sure if it is not a final version of the code then it is better to wait for the backports.

That's why I tagged @cms-sw/orp-l2 to comment

**Comment by emanueleusai** [comment]
from DQM point of view I agree with Petr. In general we prefer to test at P5 before merging the branch. If we do the opposite and new changes are required for the P5 test, we need to make sure the master is amended.
Anyway, in this case I am ok to proceed your way.

**Comment by perrotta** [comment]
> > Hi, it is somehow illogical to perform tests after commits are merged, but for sure if it is not a final version of the code then it is better to wait for the backports.
> 
> That's why I tagged @cms-sw/orp-l2 to comment

We cannot merge anything in a closed release if not tested prehentively on the master IBs and approved by the groups supposed to sign it (dqm, reconstruction, simulation, etc.). Private tests at P5 are welcome even before that, and the final evaluation of the PR will be certainly also based on them, provided the backport PR is not merged in the closed release before being fully approved.

Preparing PRs for a closed release having passed the agreed deadlines for the development one has certainly some drawback, but this is the safest way we know to avoid problems during data taking or productions planned with that closed release.

**Comment by tvami** [comment]
@clacaputo @jpata do you have any comments on the code changes from the reco point of view?

**Comment by perrotta** [comment]
> from DQM point of view I agree with Petr. In general we prefer to test at P5 before merging the branch. If we do the opposite and new changes are required for the P5 test, we need to make sure the master is amended. Anyway, in this case I am ok to proceed your way.

As I said you are welcome to test everything in the online DQM release, and this is quite helpful to everybody, Simply we will not merge in a closed release any PR which was not first fully tested in the master IBs and approved there, at least.

**Comment by jpata** [comment]
+reconstuction
- support for new GEM dataformats
- no impact on existing reco

**Comment by jpata** [comment]
> please squash at your earliest convenience!

(note that this will retrigger tests and signatures again, so probably it makes sense to request before everyone signs)

**Comment by tvami** [comment]
+1
 * resign after squash

**Comment by emanueleusai** [comment]
+1

**Comment by jpata** [comment]
+reconstuction
- support for new GEM dataformats
- no impact on existing reco
- resign since https://github.com/cms-sw/cmssw/pull/38062#issuecomment-1143742782 following a squash

**Comment by civanch** [comment]
+1

**Comment by srimanob** [comment]
+Upgrade

From coded related to the Upgrade, DataFormats is updated to the new version, `v302`. 
@yeckang Would you mind to make it clear in the PR description what is v301 and 302?

**Comment by qliphy** [comment]
> +Upgrade
> 
> From coded related to the Upgrade, DataFormats is updated to the new version, `v302`. @yeckang Would you mind to make it clear in the PR description what is v301 and 302?

@yeckang Please apply above comment.

**Comment by yeckang** [comment]
> +Upgrade
> 
> From coded related to the Upgrade, DataFormats is updated to the new version, `v302`. @yeckang Would you mind to make it clear in the PR description what is v301 and 302?

What I know is, there would be the update of firmware of AMC and OH of GEM system. And in the new firmware, we will have new format(v302) of raw data which is not fully compatible with the current version(v301).
In that reason, we should update the data format. But the unpacking module is still supporting backward compatibility for v301.

**Comment by qliphy** [comment]
+1

**Comment by tvami** [comment]
@yeckang @watson-ij please submit a merged backport of both of your PRs for 12_3_X and 12_4_X, thanks!

**Comment by tvami** [comment]
FYI: I submitted the GT change in https://github.com/cms-sw/cmssw/pull/38251

**Pull request merged**